File: Thesis/scripts/01_collection/deprecated/extract_no2_directional.js
/*
 * Extract directional NO2 climatology from Sentinel-5P.
 *
 * Paste this entire script into code.earthengine.google.com and click Run.
 * Results export to Google Drive as "no2_directional_clim".
 *
 * After export completes, download the CSV from Drive and place it at:
 *   data/stations/metadata/no2_directional_clim.csv
 *
 * Columns: stationId, direction, distance_km, month, no2_mean
 */

// ── Station coordinates (40 thesis stations) ──
// Paste your station list here. Format: [stationId, lat, lon]
// This list is loaded from station_selection_final.csv.
// Replace STATIONS_PLACEHOLDER with actual data, or load from a Fusion Table.

// Example for the first few stations — replace with full list:
var stationData = [
  // ['stationId', lat, lon],
  // Add all 40 stations from station_selection_final.csv
  // You can generate this list with:
  //   python -c "import csv; r=csv.DictReader(open('analysis/thesis_audit/station_selection_final.csv',encoding='utf-8-sig')); [print(f\"  ['{row['stationId']}', {row['lat']}, {row['lon']}],\") for row in r]"
];

if (stationData.length === 0) {
  print('ERROR: Paste station coordinates into stationData array first.');
  print('Run the Python one-liner in the comment above to generate the list.');
  // Return early — nothing to do
}

var DIRECTIONS = {
  'N': 0, 'NE': 45, 'E': 90, 'SE': 135,
  'S': 180, 'SW': 225, 'W': 270, 'NW': 315
};
var DISTANCES_KM = [5, 10, 20];
var START_DATE = '2023-01-01';
var END_DATE = '2024-12-31';

// ── Offset a point by bearing and distance ──
function offsetPoint(lat, lon, bearingDeg, distKm) {
  var R = 6371.0;
  var d = distKm / R;
  var brng = bearingDeg * Math.PI / 180;
  var lat1 = lat * Math.PI / 180;
  var lon1 = lon * Math.PI / 180;
  var lat2 = Math.asin(Math.sin(lat1) * Math.cos(d) +
                       Math.cos(lat1) * Math.sin(d) * Math.cos(brng));
  var lon2 = lon1 + Math.atan2(Math.sin(brng) * Math.sin(d) * Math.cos(lat1),
                                Math.cos(d) - Math.sin(lat1) * Math.sin(lat2));
  return [lon2 * 180 / Math.PI, lat2 * 180 / Math.PI]; // [lon, lat] for GEE
}

// ── Build sampling points ──
var pointFeatures = [];

for (var si = 0; si < stationData.length; si++) {
  var stn = stationData[si];
  var sid = String(stn[0]);
  var lat = stn[1];
  var lon = stn[2];

  // Center point
  pointFeatures.push(ee.Feature(ee.Geometry.Point([lon, lat]), {
    stationId: sid, direction: 'C', distance_km: 0
  }));

  // Directional points
  var dirNames = Object.keys(DIRECTIONS);
  for (var di = 0; di < dirNames.length; di++) {
    var dirName = dirNames[di];
    var bearing = DIRECTIONS[dirName];
    for (var dsi = 0; dsi < DISTANCES_KM.length; dsi++) {
      var dist = DISTANCES_KM[dsi];
      var offsetCoords = offsetPoint(lat, lon, bearing, dist);
      pointFeatures.push(ee.Feature(
        ee.Geometry.Point(offsetCoords), {
          stationId: sid,
          direction: dirName,
          distance_km: dist
        }
      ));
    }
  }
}

var pointsFC = ee.FeatureCollection(pointFeatures);
print('Total sampling points:', pointFeatures.length);

// ── NO2 collection with quality filter ──
var no2Collection = ee.ImageCollection('COPERNICUS/S5P/OFFL/L3_NO2')
    .filterDate(START_DATE, END_DATE)
    .filter(ee.Filter.gt('QUALITY_FLAG', 0.75))
    .select('tropospheric_NO2_column_number_density');

// ── Sample each month ──
var allResults = ee.FeatureCollection([]);

for (var month = 1; month <= 12; month++) {
  var monthlyMean = no2Collection
      .filter(ee.Filter.calendarRange(month, month, 'month'))
      .mean();

  var sampled = monthlyMean.sampleRegions({
    collection: pointsFC,
    scale: 1113.2,
    geometries: false
  });

  // Add month property
  var monthVal = month;
  var withMonth = sampled.map(function(f) {
    return f.set('month', monthVal);
  });

  allResults = allResults.merge(withMonth);
}

// ── Rename band column for clarity ──
allResults = allResults.map(function(f) {
  return ee.Feature(null, {
    stationId: f.get('stationId'),
    direction: f.get('direction'),
    distance_km: f.get('distance_km'),
    month: f.get('month'),
    no2_mean: f.get('tropospheric_NO2_column_number_density')
  });
});

print('Total result rows:', allResults.size());

// ── Export to Drive ──
Export.table.toDrive({
  collection: allResults,
  description: 'no2_directional_clim',
  fileNamePrefix: 'no2_directional_clim',
  fileFormat: 'CSV',
  selectors: ['stationId', 'direction', 'distance_km', 'month', 'no2_mean']
});

print('Export task created — click Tasks tab and click Run.');

File: Thesis/scripts/01_collection/deprecated/extract_no2_directional.py
"""
Extract directional NO2 climatology from Sentinel-5P via Google Earth Engine.

For each of 40 thesis stations, samples tropospheric NO2 column density at:
  - Center point (direction="C")
  - 8 directions (N, NE, E, SE, S, SW, W, NW) × 3 distances (5, 10, 20 km)
  = 25 points per station, 1000 points total

Computes monthly mean NO2 (Jan-Dec) across 2023-2024 with qa_value > 0.75.
Exports CSV to Google Drive.

Authentication:
  1. Install: pip install earthengine-api
  2. First run: earthengine authenticate
     (opens browser, follow prompts, paste token)
  3. Then run: python scripts/data/extract_no2_directional.py

If Python auth fails, use the JS version (extract_no2_directional.js)
in code.earthengine.google.com instead.

Output columns: stationId, direction, distance_km, month, no2_mean
Export filename on Drive: no2_directional_clim.csv
"""

import ee
import csv
import math
import os
import sys

ee.Initialize()

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
REPO_DIR = os.path.dirname(os.path.dirname(os.path.dirname(SCRIPT_DIR)))

STATION_CSV = os.path.join(REPO_DIR,
    "analysis/thesis_audit/station_selection_final.csv")

DIRECTIONS = {
    "N":  0,   "NE": 45,  "E":  90,  "SE": 135,
    "S":  180, "SW": 225, "W":  270, "NW": 315,
}
DISTANCES_KM = [5, 10, 20]
MONTHS = list(range(1, 13))
START_DATE = "2023-01-01"
END_DATE = "2024-12-31"
QA_THRESHOLD = 0.75

def offset_point(lat, lon, bearing_deg, dist_km):
    """Compute lat/lon offset by bearing and distance (Vincenty approx)."""
    R = 6371.0
    d = dist_km / R
    brng = math.radians(bearing_deg)
    lat1 = math.radians(lat)
    lon1 = math.radians(lon)
    lat2 = math.asin(math.sin(lat1) * math.cos(d) +
                     math.cos(lat1) * math.sin(d) * math.cos(brng))
    lon2 = lon1 + math.atan2(math.sin(brng) * math.sin(d) * math.cos(lat1),
                              math.cos(d) - math.sin(lat1) * math.sin(lat2))
    return math.degrees(lat2), math.degrees(lon2)

# Read stations
stations = []
with open(STATION_CSV, encoding="utf-8-sig") as f:
    reader = csv.DictReader(f)
    for row in reader:
        stations.append({
            "stationId": row["stationId"],
            "lat": float(row["lat"]),
            "lon": float(row["lon"]),
        })

print(f"Loaded {len(stations)} stations")

# Build sampling points
sample_points = []
for stn in stations:
    sid = stn["stationId"]
    lat, lon = stn["lat"], stn["lon"]
    sample_points.append({
        "stationId": sid, "direction": "C", "distance_km": 0,
        "lat": lat, "lon": lon,
    })
    for dir_name, bearing in DIRECTIONS.items():
        for dist in DISTANCES_KM:
            plat, plon = offset_point(lat, lon, bearing, dist)
            sample_points.append({
                "stationId": sid, "direction": dir_name,
                "distance_km": dist, "lat": plat, "lon": plon,
            })

print(f"Total sampling points: {len(sample_points)}")

# Build GEE feature collection of points
features = []
for i, pt in enumerate(sample_points):
    geom = ee.Geometry.Point([pt["lon"], pt["lat"]])
    props = {
        "point_id": i,
        "stationId": pt["stationId"],
        "direction": pt["direction"],
        "distance_km": pt["distance_km"],
    }
    features.append(ee.Feature(geom, props))

points_fc = ee.FeatureCollection(features)
print("Built GEE FeatureCollection")

# Process each month
no2_collection = ee.ImageCollection("COPERNICUS/S5P/OFFL/L3_NO2") \
    .filterDate(START_DATE, END_DATE) \
    .filter(ee.Filter.gt("QUALITY_FLAG", QA_THRESHOLD))

all_results = []

for month in MONTHS:
    print(f"Processing month {month}...")

    monthly = no2_collection \
        .filter(ee.Filter.calendarRange(month, month, "month")) \
        .select("tropospheric_NO2_column_number_density")

    monthly_mean = monthly.mean()

    sampled = monthly_mean.sampleRegions(
        collection=points_fc,
        scale=1113.2,  # ~0.01° ≈ S5P native resolution
        geometries=False,
    )

    sampled_list = sampled.getInfo()["features"]

    for feat in sampled_list:
        props = feat["properties"]
        no2_val = props.get("tropospheric_NO2_column_number_density", None)
        all_results.append({
            "stationId": props["stationId"],
            "direction": props["direction"],
            "distance_km": props["distance_km"],
            "month": month,
            "no2_mean": no2_val,
        })

    print(f"  Month {month}: {len(sampled_list)} samples")

# Export locally
out_path = os.path.join(REPO_DIR, "data/stations/metadata/no2_directional_clim.csv")
os.makedirs(os.path.dirname(out_path), exist_ok=True)

with open(out_path, "w", newline="", encoding="utf-8") as f:
    writer = csv.DictWriter(f, fieldnames=[
        "stationId", "direction", "distance_km", "month", "no2_mean"])
    writer.writeheader()
    writer.writerows(all_results)

print(f"\nSaved: {out_path} ({len(all_results)} rows)")
print("Done!")
